src/connector.rs

```diff
@@ -83,21 +85,28 @@ where
                 Box::pin(f)
             } else if sch == &http::uri::Scheme::HTTPS {
                 let cfg = self.tls_config.clone();
-                let hostname = dst
-                    .host()
-                    .unwrap_or_default()
-                    .to_string();
+                let hostname = match self.server_name_override.as_deref() {
+                    Some(h) => h,
+                    None => dst.host().unwrap_or_default(),
+                };
+                let hostname = match rustls::ServerName::try_from(hostname) {
+                    Ok(dnsname) => dnsname,
+                    Err(_) => {
+                        let err = io::Error::new(io::ErrorKind::Other, "invalid dnsname");
+                        let err: Box<dyn std::error::Error + Send + Sync + 'static> = Box::new(err);
+                        let res = std::future::ready(Err(err));
+                        return Box::pin(res);
+                    }
+                };
                 let connecting_future = self.http.call(dst);
 
                 let f = async move {
                     let tcp = connecting_future
                         .await
                         .map_err(Into::into)?;
                     let connector = TlsConnector::from(cfg);
-                    let dnsname = rustls::ServerName::try_from(hostname.as_str())
```

Comment:
This is unrelated change, but parsing hostname outside of the async block saves us one allocation.